### PR TITLE
Handle allocations when editing order

### PR DESCRIPTION
**Pull request by tomaszszymanski129**
I want to merge this change because it handles allocations when order line items change.

⚠️ This is a PR to feature branch.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Review by IKarbowiak — APPROVED** [review]

File: saleor/graphql/order/mutations/orders.py [hunk 1]
```diff
@@ -670,6 +671,7 @@ def perform_mutation(cls, _root, info, **data):
         cls.validate_variants(order, variants)
 
         lines = cls.add_lines_to_order(order, lines_to_add)
+        allocate_lines(lines)
```
Comment (korycins) [inline saleor/graphql/order/mutations/orders.py:674]: Correct me if I am wrong but you unify the logic for unconfirmed and draft orders. Am I right? I don't think that it is expected behaviour to allocate draft order in the same way as the unconfirmed order.

**Comment by fowczarek** [comment]
When a Staff user tries to add more items than the available dashboard display an unreadable error.
image

File: saleor/warehouse/management.py [hunk 2]
```diff
@@ -219,7 +219,21 @@ def increase_stock(
 
 
 @transaction.atomic
-def decrease_stock(order_lines_info: Iterable["OrderLineData"]):
+def increase_allocation(
+    line_info,
```
Comment (maarcingebala) [inline saleor/warehouse/management.py:223]: Missing type annotation.

File: saleor/order/utils.py [hunk 3]
```diff
@@ -367,9 +384,35 @@ def add_gift_card_to_order(order, gift_card, total_price_left):
...
+        return
+
+    if not get_order_lines_with_track_inventory([line_info]):
+        return
+
+    if old_quantity < new_quantity:
+        increase_allocation(line_info, new_quantity - old_quantity)
+    else:
+        line_info.quantity = old_quantity - new_quantity
+        decrease_allocations([line_info])
```
Comment (maarcingebala) [inline saleor/order/utils.py:407]: I'm wondering if we could unify somehow `increase_allocation` and `descrease_allocations` - one is accepting a single line, while the other is accepting a list of lines. One updates `line_info.quantity` while the other does not and we do it in line 406 manually. Shouldn't `decrease_allocations` be moved to `warehouse/management.py` where other similar functions live?

File: saleor/order/utils.py [hunk 4]
```diff
@@ -367,9 +384,35 @@ def add_gift_card_to_order(order, gift_card, total_price_left):
     return total_price_left
 
 
-def change_order_line_quantity(user, line, old_quantity, new_quantity):
+def decrease_allocations(lines_info):
+    """Decreate allocations for provided order lines."""
+    if not get_order_lines_with_track_inventory(lines_info):
+        return
+    decrease_stock(lines_info, update_stocks=False)
```
Comment (maarcingebala) [inline saleor/order/utils.py:391]: Shouldn't we decrease stock only for lines with inventory tracking?

File: saleor/warehouse/management.py [hunk 5]
```diff
@@ -266,9 +280,12 @@ def decrease_stock(order_lines_info: Iterable["OrderLineData"]):
             "quantity_allocated__sum"
         ]
 
-    _decrease_stocks_quantity(
-        order_lines_info, variant_and_warehouse_to_stock, quantity_allocation_for_stocks
-    )
+    if update_stocks:
```
Comment (maarcingebala) [inline saleor/warehouse/management.py:283]: Maybe a comment would be useful to explain in what cases we want to enter this logic?

File: saleor/warehouse/management.py [hunk 6]
```diff
@@ -219,7 +219,27 @@ def increase_stock(
 
 
 @transaction.atomic
-def decrease_stock(order_lines_info: Iterable["OrderLineData"]):
+def increase_allocation(
+    order_line,
+    quantity: int,
+):
+    """Increase allocation for order line with provided quantity."""
+    allocated = order_line.allocations.all().aggregate(Sum("quantity_allocated"))
```
[the remark this answers is not in the document]
Comment (tomaszszymanski129) [reply saleor/warehouse/management.py]: What exactly should we lock in here? After further looking into it - we might have an even bigger problem.
test case:
trigger `increase_allocation` by dashboard call and stop after delete of allocations (before it calls `allocate_stocks`)
in different shell manually run `allocate_stocks` - the shell hangs when it eventually calls bulk_create
you release the dashboard call
shell process does it's thing, but the dashboard call ends with deadlock ☠️ 
```
django.db.utils.OperationalError: deadlock detected
DETAIL:  Process 2177 waits for ShareLock on transaction 178449; blocked by process 1992.
Process 1992 waits for ShareLock on transaction 178448; blocked by process 2177.
HINT:  See server log for query details.
CONTEXT:  while locking tuple (4,127) in relation "warehouse_stock"
```

File: saleor/warehouse/management.py (hunk 5 above)
Comment (tomaszszymanski129) [reply saleor/warehouse/management.py:298]: Added a docstring about what this param does

File: saleor/order/utils.py (hunk 3 above)
Comment (tomaszszymanski129) [reply saleor/order/utils.py]: Now both accept the lines and were moved to management, should be much cleaner now.

**Review by maarcingebala — APPROVED** [review]

File: saleor/warehouse/management.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fowczarek) [reply saleor/warehouse/management.py]: We should lock all stocks for which we changing allocations.

File: saleor/order/utils.py [hunk 7]
```diff
@@ -346,6 +352,18 @@ def add_variant_to_order(order, variant, quantity, discounts=None):
             ]
         )
 
+    if line.order.is_unconfirmed():
```
Comment (korycins) [inline saleor/order/utils.py:355]: What do you think to pass the flag explicitly as a function arg?
```
def add_variant_to_order(order, variant, quantity, discounts=None, allocate_stock=False):
```
IMO it should be more flexible than the current flow.

Comment (tomaszszymanski129) [reply saleor/order/utils.py:355]: Yep, sounds good 👍🏻

**Review by fowczarek — APPROVED** [review]

**Pull request merged**